basictracer/propagation_ls.go

```diff
@@ -0,0 +1,105 @@
+package basictracer
+
+import (
+	"encoding/base64"
+
+	"github.com/gogo/protobuf/proto"
+	lightstep "github.com/lightstep/lightstep-tracer-go/lightsteppb"
+	opentracing "github.com/opentracing/opentracing-go"
+)
+
+type lightstepBinaryPropagator struct{}
+
+type lightstepBinaryCarrier struct{}
+
+var BinaryCarrier lightstepBinaryCarrier
+
+func (_ *lightstepBinaryPropagator) Inject(
+	spanContext opentracing.SpanContext,
+	opaqueCarrier interface{},
+) error {
+	sc, ok := spanContext.(SpanContext)
+	if !ok {
+		return opentracing.ErrInvalidSpanContext
+	}
+	var scarrier *string
+	var bcarrier *[]byte
+	switch t := opaqueCarrier.(type) {
+	case *string:
+		scarrier = t
+	case *[]byte:
+		bcarrier = t
+	default:
+		return opentracing.ErrInvalidCarrier
+	}
+	pb := &lightstep.BinaryCarrier{}
+	pb.BasicCtx = &lightstep.BasicTracerCarrier{
+		TraceId:      sc.TraceID,
+		SpanId:       sc.SpanID,
+		Sampled:      true,
+		BaggageItems: sc.Baggage,
+	}
+	data, err := proto.Marshal(pb)
+	if err != nil {
+		return err
+	}
+	if bcarrier == nil {
+		*scarrier = base64.StdEncoding.EncodeToString(data)
+	} else {
+		*bcarrier = make([]byte, base64.StdEncoding.EncodedLen(len(data)))
+		base64.StdEncoding.Encode(*bcarrier, data)
+	}
+	return nil
+}
```

Comment:
Small readability/optimization improvement: please consider this version of `Inject` that has fewer conditionals and extra allocations (and something similar for `Extract`).
```go
{
	sc, ok := spanContext.(SpanContext)
	if !ok {
		return opentracing.ErrInvalidSpanContext
	}

	data, err := proto.Marshal(&lightstep.BinaryCarrier{
		BasicCtx: &lightstep.BasicTracerCarrier{
			TraceId:      sc.TraceID,
			SpanId:       sc.SpanID,
			Sampled:      true,
			BaggageItems: sc.Baggage,
		}
	})
	if err != nil {
		return err
	}

	switch carrier := opaqueCarrier.(type) {
	case *string:
		*carrier = base64.StdEncoding.EncodeToString(data)
	case *[]byte:
		*carrier = make([]byte, base64.StdEncoding.EncodedLen(len(data)))
		base64.StdEncoding.Encode(*carrier, data)
	default:
		return opentracing.ErrInvalidCarrier
	}

	return nil
}
```